text/0000-implied-bounds.md

```diff
@@ -0,0 +1,739 @@
+
+    type Assoc = AssocTyValue;
+
+    /* ... */
+}
+```
+We rely on the following assumptions inside the body of the impl:
+* `Expanded({ WhereClause1, ..., WhereClausek })`
+* `WF(T)` for all `T ∈ Union(InputTypes(T0), InputTypes(T1), ..., InputTypes(Tn))`
+* `WF(Xi)` for all `i`
+
+Based on these assumptions, the impl declaration has to prove `WF(T0: Trait<r'..., T1, ..., Tn>)` and `WF(T)` for all `T ∈ InputTypes(AssocTyValue)`. Note that associated fns can be seen as (higher-kinded) associated types, but since fn pointers are always well-formed and do not carry input types other than themselves, this is fine.
+
+Associated fns make their normal assumptions + the set of assumptions made by the impl. Things to prove inside associated fns do not differ from normal fns.
+
+Note that when projecting out of a type, one must automatically prove that the trait reference holds because of the `WfProjection` rule.
+
+Examples:
+
+```rust
+struct Set<K: Hash> { ... }
+
+trait Foo where Self: Clone {
+    fn foo();
+}
+
+fn only_hash<T: Hash>() { }
+
+impl<K: Clone> Foo for Set<K> {
+    // Inside here: we assume `WF(Set<K>)`, `K: Clone`, `WF(K: Clone)`, `WF(K)`.
+    // Also, we must prove `WF(Set<K>: Foo)`.
+
+    fn foo() {
+        only_hash::<K>()
+    }
+}
+```
+
+```rust
+struct Set<K: Hash> { ... }
+
+trait Foo {
+    type Item;
+}
+
+// We need an explicit `K: Hash` bound in order to prove that the associated type value `Set<K>` is WF.
+impl<K: Hash> Foo for K {
+    type Item = Set<K>;
+}
+```
+
+```rust
+trait Foo {
+    type Item;
+}
+
+impl<T> Foo for T where T: Clone {
+    type Item = f32;
+}
+
+fn foo<T: Foo>(arg: T) {
+    // We must prove `WF(<T as Foo>::Item)` hence prove that `T: Foo`: ok this is in our assumptions.
+    let a = <T as Foo>::Item;
+}
+
+fn bar<T: Clone>(arg: T) {
+    // We must prove `WF(<T as Foo>::Item)` hence prove that `T: Foo`: ok, use the impl.
+    let a = <T as Foo>::Item;
+}
+```
+
+### **Inherent impls**
+Given an inherent impl:
+```rust
+impl<r..., X1, ..., Xn> SelfTy where WhereClause1, ..., WhereClausek {
+    /* body of the impl inside here */
+}
+```
+We rely on the following assumptions inside the body of the impl:
+* `Expanded({ WhereClause1, ..., WhereClausek })`
+* `WF(T)` for all `T ∈ InputTypes(SelfTy)`
+* `WF(Xi)` for all `i`
+
+Methods make their normal assumptions + the set of assumptions made by the impl. Things to prove inside methods do not differ from normal fns.
+
+A caller of a method has to prove that the where clauses defined on the impl hold, in addition to the requirements for calling general fns.
+
+## Proving well-formedness for input types
+[proving-wf-input-types]: #proving-well-formedness-for-input-types
+
+One would have noticed that we only prove well-formedness for input types in a lazy way (e.g., inside function bodies). This means that if we have a function:
+```rust
+struct Set<K: Hash> { ... }
+struct NotHash;
+
+fn foo(arg: Set<NotHash>) { ... }
```

Comment:
That's not correct in practice, because

1. rustc will try to prove that `ENV(foo) |- WF(Set<NotHash>)`, because of one of the "the type of everything within a function must be well-formed" rules.
2. we theoretically have `WF(Set<NotHash>)` in our environment. Practically, we don't - the current implementation:
2.A. does not allow using environment predicates that aren't trait predicates, projection predicates, or "ParamWf" predicates. This is IIRC not done for a good reason, and is not needed.
2.B. ignores environment predicates that do not contain any parameters. This is done because these predicates either hold (and are therefore unneeded) or don't hold (and therefore can't be used in a valid program). This helps caching.
2.C. has "anti-getting-lost" rules. The rules "as written" mean, that if one has a trait
    ```Rust
    trait Foo { type Bar: Sized; }
    ```
    Then there is a trait-system rule
    ```
    τ type
    WF(τ)
    τ: Foo
    ----------------
    <τ as Foo>::Bar: Sized
    ```
    Then in order to prove that there is no impl for `[T]: Sized`, one has to check that we don't have a type `[T] = <some-type as Foo>::Bar`. That can require going over all the impls to see that there is no applicable impl, so the current implementation in rustc does not use the trait-system rule unless it sees that the type is already a projection type.